File: landxml_parser.py
# title: landxml parser
# author: kang taewook
# email: laputa99999@gmail.com
# description: convert landxml to json
# 
import os, sys, re, json, math, traceback, pandas as pd, numpy as np, numpy as np
import xml.etree.ElementTree as elemTree
from nltk.tokenize import WhitespaceTokenizer

sys.path.insert(0, os.path.dirname(os.getcwd()))

def import_json(fname): 
	data = None
	with open(fname) as f:
		data = json.load(f)
	return data

def get_xml_tag_name(tag):
	index = tag.find("}") 
	if index >= 0:
		tag = tag[index + 1: len(tag)]

	return tag

def get_key_in_dict(dict_data, index):
	list = [i for i in dict_data.keys()]
	if len(list) <= index:
		return None
	return list[index]
	
class XML_parser:
	_find = 0
	_nodes = []
	
	def clear_finall_node(self):
		self._nodes.clear()
		
	def add_findall_node(self, node):
		for n in self._nodes:
			if n == node:
				return
		self._nodes.append(node)
		return

	def is_match_attrib(self, node, attr):
		i = 0
		for key in attr:
			try:
				# print("cond: " + key)
				value = attr[key]
				if node.attrib[key].find(value) >= 0: 
					continue
				return False
			except:
				return False
			i = i + 1
		print("* matched. " + str(node.attrib))
		return True

	def findall(self, node, tag, attr = ""):
		#tag_cond = {'base_tag': 'CrossSect', 'name': "0+040.00"}
		for child in node:
			try:
				if child.tag.find(tag) >= 0:
					# print(str(find) + ".tag: " + child.tag + ": " + str(child.attrib))
					if self.is_match_attrib(child, attr):
						self.add_findall_node(child)
						return child					
			except ValueError:
				pass
			result = self.findall(child, tag, attr)
			if result != None:
				self.add_findall_node(child)
				return result
			self._find = self._find + 1
		return None

class landxml:
	_model_data = None
	def __init__(slef):
		pass

	def load(self, fpath):
		try:
			tree = elemTree.parse(fpath)
			root = tree.getroot()        

			model = self.parsing(root)
			self._model_data = model
			return model
		except Exception as e:
			traceback.print_exc()
			pass
		return None

	def save(self, fpath):
		try:
			if self._model_data == None:
				return
			
			with open(fpath, 'w') as json_file:
				json.dump(self._model_data, json_file, indent = 2)
		except Exception as e:
			traceback.print_exc()
			pass

	def get_points_in_text(self, text):
		tk = WhitespaceTokenizer()
		tokens = tk.tokenize(text)
		pcd = []
		i = 0
		count = round(len(tokens) / 2)
		for i in range(count):
			x = float(tokens[i * 2])
			y = float(tokens[i * 2 + 1])
			pcd.append((x, y))
		return pcd

	def get_attrib_text(self, node):
		tag = get_xml_tag_name(node.tag)
		print(tag, node.attrib)
		data = {}
		data[tag] = {}
		data[tag]['attrib'] = node.attrib
		if node.text == None:
			data[tag]['text'] = ""
		else:
			data[tag]['text'] = node.text.replace('\t', '')
			data[tag]['text'] = data[tag]['text'].replace('\n', '')
		data[tag]['list'] = []

		return tag, node.attrib, node.text, data
	
	def parsing_curve(self, node, model):
		for child in node:
			tag, attrib, text, data = self.get_attrib_text(child)
			if text != None:
				data[tag]['points'] = self.get_points_in_text(text)
			model.append(data)

			if tag.find("Start") >= 0:
				pass
			elif tag.find("Center") >= 0:
				pass
			elif tag.find("End") >= 0:
				pass
			elif tag.find("PI") >= 0:
				pass

		return model

	def parsing_curve_set(self, node, model):
		for child in node:
			tag, attrib, text, data = self.get_attrib_text(child)

			if child.tag.find("Line") >= 0 or child.tag.find("Curve") >= 0:
				model.append(data)
				self.parsing_curve(child, data[tag]['list'])

		return model

	def parsing_cross_sect_surf(self, node, model):
		for child in node:
			tag, attrib, text, data = self.get_attrib_text(child)

			if child.tag.find("PntList2D") < 0:
				continue

			data[tag]['points'] = self.get_points_in_text(text)
			model.append(data)

		return model

	def parsing_design_cross_sect_surf(self, node, model):
		for child in node:
			tag, attrib, text, data = self.get_attrib_text(child)

			if child.tag.find("CrossSectPnt") < 0:
				continue

			data[tag]['points'] = self.get_points_in_text(text)	# TBD
			model.append(data)

		return model

	def parsing_cross_sect(self, node, model):
		for child in node:
			tag, attrib, text, data = self.get_attrib_text(child)

			if child.tag.find("DesignCrossSectSurf") >= 0:
				model.append(data)
				self.parsing_design_cross_sect_surf(child, data[tag]['list'])
			elif child.tag.find("CrossSectSurf") >= 0:
				model.append(data)
				self.parsing_cross_sect_surf(child, data[tag]['list'])

		return model

	def parsing_cross_sects(self, node, model):
		for child in node:
			tag, attrib, text, data = self.get_attrib_text(child)

			if child.tag.find("CrossSect") >= 0:
				model.append(data)
				self.parsing_cross_sect(child, data[tag]['list'])

		return model

	def parsing_prof_surf(self, node, model):
		for child in node:
			tag, attrib, text, data = self.get_attrib_text(child)

			if child.tag.find("PntList2D") >= 0:
				data[tag]['points'] = self.get_points_in_text(text)
				model.append(data)

		return model

	def parsing_prof_align(self, node, model):
		for child in node:
			tag, attrib, text, data = self.get_attrib_text(child)

			if child.tag.find("PVI") >= 0:
				data[tag]['points'] = self.get_points_in_text(text)
				model.append(data)
			elif child.tag.find("ParaCurve") >= 0:
				data[tag]['points'] = self.get_points_in_text(text)
				model.append(data)

		return model

	def parsing_profile(self, node, model):
		for child in node:
			tag, attrib, text, data = self.get_attrib_text(child)

			if child.tag.find("ProfSurf") >= 0:
				model.append(data)
				self.parsing_prof_surf(child, data[tag]['list'])
			elif child.tag.find("ProfAlign") >= 0:
				model.append(data)
				self.parsing_prof_align(child, data[tag]['list'])

		return model

	def parsing_coordgeom(self, node, model):
		for child in node:
			tag, attrib, text, data = self.get_attrib_text(child)

			if tag.find("CoordGeom") >= 0:
				model.append(data)
				self.parsing_curve_set(child, data[tag]['list'])
			elif tag.find("CrossSects") >= 0:
				model.append(data)
				self.parsing_cross_sects(child, data[tag]['list'])
			elif tag.find("Profile") >= 0:
				model.append(data)
				self.parsing_profile(child, data[tag]['list'])

		return model

	def parsing_alignment(self, node, model):
		for child in node:
			tag, attrib, text, data = self.get_attrib_text(child)

			if tag.find("Alignment") >= 0:
				model.append(data)
				self.parsing_coordgeom(child, data[tag]['list'])	

		return model

	def parsing(self, node):
		model = []

		try:
			for child in node:
				tag, attrib, text, data = self.get_attrib_text(child)

				if tag.find("Project") >= 0:
					pass   
				elif tag.find("Units") >= 0:
					pass   
				elif tag.find("Application") >= 0:
					pass
				elif tag.find("Alignments") >= 0:
					model.append(data)
					self.parsing_alignment(child, data[tag]['list'])
				elif tag.find("Roadways") >= 0:
					pass   
				elif tag.find("Surfaces") >= 0:
					pass   
		except Exception as e:
			traceback.print_exc()
			pass			
		return model
	
def test():
	lxml = landxml()
	model = lxml.load('sample.xml')
	print(model)
	lxml.save('output.json')

if __name__ == "__main__":
	test()

File: javascript/landxml_parser.js
const fs = require('fs');
const path = require('path');
const natural = require('natural');
const tokenizer = new natural.WordTokenizer();
const { parseString } = require('xml2js');

function importJson(fname) {
    let data = null;
    try {
        data = JSON.parse(fs.readFileSync(fname, 'utf8'));
    } catch (err) {
        console.error(err);
    }
    return data;
}

function getXmlTagName(tag) {
    const index = tag.indexOf("}");
    if (index >= 0) {
        tag = tag.substring(index + 1);
    }
    return tag;
}

function getKeyInDict(dictData, index) {
    const keys = Object.keys(dictData);
    if (keys.length <= index) {
        return null;
    }
    return keys[index];
}

class XMLParser {
    constructor() {
        this._find = 0;
        this._nodes = [];
    }

    clearFinalNode() {
        this._nodes = [];
    }

    addFindAllNode(node) {
        for (const n of this._nodes) {
            if (n === node) {
                return;
            }
        }
        this._nodes.push(node);
    }

    isMatchAttrib(node, attr) {
        let i = 0;
        for (const key in attr) {
            try {
                const value = attr[key];
                if (node.attrib[key].indexOf(value) >= 0) {
                    continue;
                }
                return false;
            } catch {
                return false;
            }
            i++;
        }
        console.log("* matched. " + JSON.stringify(node.attrib));
        return true;
    }

    findAll(node, tag, attr = "") {
        for (const child of node.children) {
            try {
                if (child.tag.indexOf(tag) >= 0) {
                    if (this.isMatchAttrib(child, attr)) {
                        this.addFindAllNode(child);
                        return child;
                    }
                }
            } catch (error) {
                // Handle error if necessary
            }
            const result = this.findAll(child, tag, attr);
            if (result !== null) {
                this.addFindAllNode(child);
                return result;
            }
            this._find++;
        }
        return null;
    }
}

function loadXml(fpath) {
    return new Promise((resolve, reject) => {
        fs.readFile(fpath, 'utf-8', (err, data) => {
            if (err) {
                reject(err);
                return;
            }
            parseString(data, (err, result) => {
                if (err) {
                    reject(err);
                    return;
                }
                resolve(result);
            });
        });
    });
}

class LandXML {
    constructor() {
        this._modelData = null;
    }

    async load(fpath) {
        try {
            const xmlData = await loadXml(fpath);

            const model = this.parsing(xmlData);
            this._modelData = model;
            return model;
        } catch (error) {
            console.error(error);
        }
        return null;
    }

    getModelsData() {
        return this._modelData;
    }

    save(fpath) {
        try {
            if (this._modelData === null) {
                return false;
            }

            fs.writeFileSync(fpath, JSON.stringify(this._modelData, null, 2));
        } catch (error) {
            console.error(error);
        }
        return true;
    }

    getPointsInText(text) {
        const tokens = tokenizer.tokenize(text);
        const pcd = [];
        for (let i = 0; i < Math.round(tokens.length / 2); i++) {
            const x = parseFloat(tokens[i * 2]);
            const y = parseFloat(tokens[i * 2 + 1]);
            pcd.push([x, y]);
        }
        return pcd;
    }

    getAttribText(node) {
        const tag = this.getXmlTagName(node.tag);
        console.log(tag, node.attrib);
        const data = {};
        data[tag] = {};
        data[tag]['attrib'] = node.attrib;
        if (node.text === null) {
            data[tag]['text'] = "";
        } else {
            data[tag]['text'] = node.text.replace('\t', '').replace('\n', '');
        }
        data[tag]['list'] = [];

        return [tag, node.attrib, node.text, data];
    }

    parsingCurve(node, model) {
        for (const child of node.children) {
            const [tag, attrib, text, data] = this.getAttribText(child);
            if (text !== null) {
                data[tag]['points'] = this.getPointsInText(text);
            }
            model.push(data);

            // Handling different tags, if necessary
            // if (tag.indexOf("Start") >= 0) {}
            // else if (tag.indexOf("Center") >= 0) {}
            // else if (tag.indexOf("End") >= 0) {}
            // else if (tag.indexOf("PI") >= 0) {}
        }

        return model;
    }

    parsingCurveSet(node, model) {
        for (const child of node.children) {
            const [tag, attrib, text, data] = this.getAttribText(child);

            if (child.tag.indexOf("Line") >= 0 || child.tag.indexOf("Curve") >= 0) {
                model.push(data);
                this.parsingCurve(child, data[tag]['list']);
            }
        }

        return model;
    }

    parsingCrossSectSurf(node, model) {
        for (const child of node.children) {
            const [tag, attrib, text, data] = this.getAttribText(child);

            if (child.tag.indexOf("PntList2D") >= 0) {
                data[tag]['points'] = this.getPointsInText(text);
                model.push(data);
            }
        }

        return model;
    }

    parsingDesignCrossSectSurf(node, model) {
        for (const child of node.children) {
            const [tag, attrib, text, data] = this.getAttribText(child);

            if (child.tag.indexOf("CrossSectPnt") >= 0) {
                data[tag]['points'] = this.getPointsInText(text); // TBD
                model.push(data);
            }
        }

        return model;
    }

    parsingCrossSect(node, model) {
        for (const child of node.children) {
            const [tag, attrib, text, data] = this.getAttribText(child);

            if (child.tag.indexOf("DesignCrossSectSurf") >= 0) {
                model.push(data);
                this.parsingDesignCrossSectSurf(child, data[tag]['list']);
            } else if (child.tag.indexOf("CrossSectSurf") >= 0) {
                model.push(data);
                this.parsingCrossSectSurf(child, data[tag]['list']);
            }
        }

        return model;
    }

    parsingCrossSects(node, model) {
        for (const child of node.children) {
            const [tag, attrib, text, data] = this.getAttribText(child);

            if (child.tag.indexOf("CrossSect") >= 0) {
                model.push(data);
                this.parsingCrossSect(child, data[tag]['list']);
            }
        }

        return model;
    }

    parsingProfSurf(node, model) {
        for (const child of node.children) {
            const [tag, attrib, text, data] = this.getAttribText(child);

            if (child.tag.indexOf("PntList2D") >= 0) {
                data[tag]['points'] = this.getPointsInText(text);
                model.push(data);
            }
        }

        return model;
    }

    parsingProfAlign(node, model) {
        for (const child of node.children) {
            const [tag, attrib, text, data] = this.getAttribText(child);

            if (child.tag.indexOf("PVI") >= 0 || child.tag.indexOf("ParaCurve") >= 0) {
                data[tag]['points'] = this.getPointsInText(text);
                model.push(data);
            }
        }

        return model;
    }

    parsingProfile(node, model) {
        for (const child of node.children) {
            const [tag, attrib, text, data] = this.getAttribText(child);

            if (child.tag.indexOf("ProfSurf") >= 0) {
                model.push(data);
                this.parsingProfSurf(child, data[tag]['list']);
            } else if (child.tag.indexOf("ProfAlign") >= 0) {
                model.push(data);
                this.parsingProfAlign(child, data[tag]['list']);
            }
        }

        return model;
    }

    parsingCoordGeom(node, model) {
        for (const child of node.children) {
            const [tag, attrib, text, data] = this.getAttribText(child);

            if (tag.indexOf("CoordGeom") >= 0) {
                model.push(data);
                this.parsingCurveSet(child, data[tag]['list']);
            } else if (tag.indexOf("CrossSects") >= 0) {
                model.push(data);
                this.parsingCrossSects(child, data[tag]['list']);
            } else if (tag.indexOf("Profile") >= 0) {
                model.push(data);
                this.parsingProfile(child, data[tag]['list']);
            }
        }

        return model;
    }

    parsingAlignment(node, model) {
        for (const child of node.children) {
            const [tag, attrib, text, data] = this.getAttribText(child);

            if (tag.indexOf("Alignment") >= 0) {
                model.push(data);
                this.parsingCoordGeom(child, data[tag]['list']);
            }
        }

        return model;
    }

    parsing(node) {
        const model = [];

        try {
            for (const child of node.children) {
                const [tag, attrib, text, data] = this.getAttribText(child);

                if (tag.indexOf("Project") >= 0) {
                    // Handle Project parsing
                } else if (tag.indexOf("Units") >= 0) {
                    // Handle Units parsing
                } else if (tag.indexOf("Application") >= 0) {
                    // Handle Application parsing
                } else if (tag.indexOf("Alignments") >= 0) {
                    model.push(data);
                    this.parsingAlignment(child, data[tag]['list']);
                } else if (tag.indexOf("Roadways") >= 0) {
                    // Handle Roadways parsing
                } else if (tag.indexOf("Surfaces") >= 0) {
                    // Handle Surfaces parsing
                }
            }
        } catch (error) {
            console.error(error);
        }

        return model;
    }
}

// Define test function
function test() {
    const parser = new LandXML();
    // Load landxml file
    // const model = parser.load('landxml_railway_sample.xml'); // Load railway landxml file
    const model = parser.load('./landxml_road_sample.xml'); // Load road landxml file
    console.log(model);
    // Save model data to JSON file
    parser.save('output_landxml.json');
}

// Check if the script is run directly
if (require.main === module) {
    test();
}


